core/src/test/java/org/apache/iceberg/io/InMemoryFileIO.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.io;
+
+import java.util.Map;
+import org.apache.iceberg.exceptions.NotFoundException;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+
+public class InMemoryFileIO implements FileIO {
```

Comment:
I would probably consider introducing this as a separate PR. That way it can be reviewed independently from the Snowflake Catalog PR

Reply:
Yeah, started out with this just being in the snowflake package; I expanded this and moved it here due to https://github.com/apache/iceberg/pull/6428#discussion_r1051680729

I could move the expanded functionality back into snowflake package and then send a future followup just moving the file back into the shared package.

@rdblue what do you think?

Reply:
I think @nastra was suggesting that we just take the in-memory implementation as a separate PR and get that committed before this PR, not change the packaging and follow up with a PR to move it.  Then you can just rebase this PR on top.

Reply:
Ah makes sense, just split out https://github.com/apache/iceberg/pull/6538